classes/user_exception_selector.php

```diff
@@ -0,0 +1,193 @@
+        $countsql = "SELECT COUNT(1)
+                       FROM {user} u
+                       JOIN {user_enrolments} AS ue ON ue.userid = u.id
+                       JOIN {enrol} AS en ON ue.enrolid = en.id
+                       JOIN {course} AS course ON en.courseid = course.id
+                     $ejoin {availability_shibboleth2fa_e} e ON (e.userid = u.id AND e.courseid = course.id AND e.skipauth = 1)
+                      WHERE $wherecondition
+                            AND course.id = :courseid";
+        if (!$this->is_validating()) {
+            $userscount = $DB->count_records_sql($countsql, $params);
+            if ($userscount > $this->maxusersperpage) {
+                return $this->too_many_results($search, $userscount);
+            }
+        }
+        $sql = preg_replace(
+            pattern: '/^SELECT COUNT\(1\)/',
+            replacement: "SELECT {$this->required_fields_sql('u')}",
+            subject: $countsql,
+            limit: 1,
+        );
+        [$sort, $sortparams] = users_order_by_sql('u', $search, $this->accesscontext);
+        $users = $DB->get_records_sql("$sql ORDER BY $sort", $params + $sortparams);
+        if (empty($users)) {
+            return [];
+        }
+        if ($search) {
+            $groupname = get_string("{$stringkey}_matching", 'availability_shibboleth2fa', $search);
+        } else {
+            $groupname = get_string($stringkey, 'availability_shibboleth2fa');
+        }
+        return [$groupname => $users];
+    }
+
+    /**
+     * Returns the options needed to recreate the given {@see user_exception_selector}.
+     *
+     * @return array Options to be passed to the class' constructor.
+     */
+    protected function get_options(): array {
+        return parent::get_options() + [
+            'courseid' => $this->courseid,
+            'skipauth' => $this->skipauth,
+        ];
+    }
+
+    /**
+     * If any users have been selected, an exception for them is added/removed in the associated course.
+     *
+     * If the selector was created for users that are already exempt from 2FA requirements in the course (`skipauth` set to `true`),
+     * the exceptions will be removed for the selected users. If it was created for users that are not (yet) exempt from 2FA
+     * requirements in the course (`skipauth` set to `false`), exceptions will be added for the selected users.
+     *
+     * Only performs the modifications, if {@see confirm_sesskey} returns `true`.
+     *
+     * Calls the {@see invalidate_selected_users} method at the end.
+     *
+     * @return object[] The selected users for whom the exception status has been modified.
+     * @throws dml_exception
+     */
+    public function set_exceptions_for_selected_users(): array {
+        $users = $this->get_selected_users();
+        if (empty($users) || !confirm_sesskey()) {
```

Comment:
I would prefer that you check `confirm_sesskey()` before calling this function. The name of the function does not indicate that it validates access.

Reply:
If I just pull it out of that method and call it everytime in `manage.php`, that would cause an error, when there is no `sesskey` at all, which is the case, when the form page is simply `GET` requested. I guess I could just check for that request method:

```php
if ($_SERVER['REQUEST_METHOD'] === 'POST' && confirm_sesskey()) {
    $potentialuserselector->set_exceptions_for_selected_users();
    $currentuserselector->set_exceptions_for_selected_users();
}
```

What do you think?

Reply:
I prefer this version :)

Reply:
I re-introduced the previous checks for which of the two submit buttons was pressed to improve the user experience, making the HTTP method check unnecessary.